From 9c85f509761b3f5484f988c485bc42b555c03871 Mon Sep 17 00:00:00 2001
From: Bastien Nocera <hadess@hadess.net>
Date: Sat, 29 Nov 2008 16:17:53 +0000
Subject: [PATCH] Add icon for other audio device

.
---
 src/dbus-hci.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/src/dbus-hci.c b/src/dbus-hci.c
index f2b6f2e..a0de9a0 100644
--- a/src/dbus-hci.c
+++ b/src/dbus-hci.c
@@ -85,6 +85,8 @@ const char *class_to_icon(uint32_t class)
 			return "audio-card";	/* Headset */
 		case 0x06:
 			return "audio-card";	/* Headphone */
+		default:
+			return "audio-card";	/* Other audio device */
 		}
 		break;
 	case 0x05:
-- 
1.6.0.4

