--- bluez-4.42/tools/hid2hci.c.old	2009-11-02 17:13:56.000000000 +0000
+++ bluez-4.42/tools/hid2hci.c	2009-11-02 17:24:39.000000000 +0000
@@ -25,6 +25,7 @@
 #include <errno.h>
 #include <fcntl.h>
 #include <stdint.h>
+#define _GNU_SOURCE 1
 #include <string.h>
 #include <getopt.h>
 #include <sys/ioctl.h>
@@ -32,7 +33,63 @@
 #include <usb.h>
 
 #include "libudev.h"
-#include "libudev-private.h"
+
+/*
+ * Concatenates strings. In any case, terminates in _all_ cases with '\0'
+ * and moves the @dest pointer forward to the added '\0'. Returns the
+ * remaining size, and 0 if the string was truncated.
+ */
+size_t util_strpcpy(char **dest, size_t size, const char *src)
+{
+	size_t len;
+
+	len = strlen(src);
+	if (len >= size) {
+		if (size > 1)
+			*dest = mempcpy(*dest, src, size-1);
+		size = 0;
+		*dest[0] = '\0';
+	} else {
+		if (len > 0) {
+			*dest = mempcpy(*dest, src, len);
+			size -= len;
+		}
+		*dest[0] = '\0';
+	}
+	return size;
+}
+
+/* concatenates list of strings, moves dest forward */
+size_t util_strpcpyl(char **dest, size_t size, const char *src, ...)
+{
+	va_list va;
+
+	va_start(va, src);
+	do {
+		size = util_strpcpy(dest, size, src);
+		src = va_arg(va, char *);
+	} while (src != NULL);
+	va_end(va);
+
+	return size;
+}
+
+/* concatenates list of strings */
+size_t util_strscpyl(char *dest, size_t size, const char *src, ...)
+{
+	va_list va;
+	char *s;
+
+	va_start(va, src);
+	s = dest;
+	do {
+		size = util_strpcpy(&s, size, src);
+		src = va_arg(va, char *);
+	} while (src != NULL);
+	va_end(va);
+
+	return size;
+}
 
 enum mode {
 	HCI = 0,
--- bluez-4.42/scripts/bluetooth-hid2hci.rules.old	2009-11-03 09:51:29.000000000 +0000
+++ bluez-4.42/scripts/bluetooth-hid2hci.rules	2009-11-03 09:51:47.000000000 +0000
@@ -20,7 +20,7 @@ ENV{DEVTYPE}!="usb_device", GOTO="hid2hc
 # Unfortunately the only event seen is the BT device disappearing, so the mouse
 # device needs to be chased down on the USB bus.
 ATTR{bDeviceClass}=="e0", ATTR{bDeviceSubClass}=="01", ATTR{bDeviceProtocol}=="01", ATTR{idVendor}=="413c", \
-  ENV{REMOVE_CMD}="/sbin/udevadm trigger --action=change --subsystem-match=usb --property-match=HID2HCI_SWITCH=1"
+  ENV{REMOVE_CMD}="/sbin/udevadm trigger --action=change --subsystem-match=usb"
 
 # CSR devices
 ATTR{idVendor}=="0a12|0458|05ac", ATTR{idProduct}=="1000", RUN+="hid2hci --method=csr --devpath=%p"
